drivers/gpio/gpio_sim.c

```diff
@@ -0,0 +1,639 @@
+		(struct gpio_sim_data *)port->driver_data;
+	const struct gpio_sim_config *const config =
+		(const struct gpio_sim_config *)port->config_info;
+
+	for (i = 0, *interrupts = 0; mask && i < config->num_pins;
+	     ++i, mask >>= 1, prev_values >>= 1, values >>= 1) {
+		if ((mask & 1) == 0) {
+			continue;
+		}
+
+		prev_bit = ((prev_values & 1) != 0);
+		bit = ((values & 1) != 0);
+
+		switch (drv_data->flags[i] & GPIO_SIM_INT_BITMASK) {
+		case GPIO_INT_EDGE_RISING:
+			if (detect_edge && !prev_bit && bit) {
+				*interrupts |= BIT(i);
+			}
+			break;
+		case GPIO_INT_EDGE_FALLING:
+			if (detect_edge && prev_bit && !bit) {
+				*interrupts |= BIT(i);
+			}
+			break;
+		case GPIO_INT_EDGE_BOTH:
+			if (detect_edge && prev_bit != bit) {
+				*interrupts |= BIT(i);
+			}
+			break;
+		case GPIO_INT_LEVEL_LOW:
+			if (!bit) {
+				*interrupts |= BIT(i);
+			}
+			break;
+		case GPIO_INT_LEVEL_HIGH:
+			if (bit) {
+				*interrupts |= BIT(i);
+			}
+			break;
+		case 0:
+		case GPIO_INT_DISABLE:
+			break;
+		default:
+			LOG_DBG("unhandled case %u",
+				drv_data->flags[i] & GPIO_SIM_INT_BITMASK);
+			break;
+		}
+	}
+}
+
+/**
+ * @brief Trigger possible interrupt events after an input pin has changed
+ *
+ * For more information, see @ref gpio_interface.
+ *
+ * The caller must ensure that @ref gpio_sim_data.mu is locked.
+ *
+ * @param port The simulated GPIO port
+ * @param mask The mask of pins that have changed
+ * @param prev_values Previous pin values
+ * @param values Current pin values
+ */
+static void gpio_sim_pend_interrupt(struct device *port, gpio_port_pins_t mask,
+				    gpio_port_value_t prev_values,
+				    gpio_port_value_t values)
+{
+	gpio_port_pins_t interrupts;
+	struct gpio_sim_data *const drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+
+	for (gpio_sim_gen_interrupt_bits(port, mask, prev_values, values,
+					 &interrupts, true);
+	     interrupts; gpio_sim_gen_interrupt_bits(
+		     port, mask, prev_values, values, &interrupts, false)) {
+		gpio_fire_callbacks(&drv_data->callbacks, port, interrupts);
+	}
+}
+
+/* documented in drivers/gpio/gpio_sim.h */
+int gpio_sim_input_set_masked(struct device *port, gpio_port_pins_t mask,
+			      gpio_port_value_t values)
+{
+	int ret;
+	gpio_port_pins_t input_mask;
+	gpio_port_pins_t prev_values;
+	struct gpio_sim_data *const drv_data =
+		(struct gpio_sim_data *)port->driver_data;
+	const struct gpio_sim_config *const config =
+		(const struct gpio_sim_config *)port->config_info;
+
+	if (mask == 0) {
+		return 0;
+	}
+
+	if (~config->common.port_pin_mask & mask) {
+		return -EINVAL;
+	}
+
+	if (values & ~mask) {
+		return -EINVAL;
+	}
+
+	k_mutex_lock(&drv_data->mu, K_FOREVER);
+
+	input_mask = gpio_sim_input_mask(port);
+	if (~input_mask & mask) {
+		ret = -EINVAL;
+		goto unlock;
+	}
+
+	prev_values = drv_data->vals;
+	drv_data->vals &= ~mask;
+	drv_data->vals |= values;
+	values = drv_data->vals;
+
+	gpio_sim_pend_interrupt(port, mask, prev_values, values);
```

Comment:
Actually I just remembered (perhaps incorrectly) that Zephyr's `k_mutex` is recursive, in which case this could work fine.

It's still pretty unusual to use `k_mutex` instead of `k_sem` for driver serial access, though I'm trying to get buy-in to change that.